### PR TITLE
[inductor] support masked_scatter w/ unbacked sized source

### PR DESCRIPTION
**Pull request by ColinPeppler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138083

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -3926,7 +3926,9 @@ def _unsafe_masked_index(x, mask, indices, fill):
         lambda: "tensors used as masks must be bool tensors",
     )
 
-    if x.numel() == 0:
+    from torch.fx.experimental.symbolic_shapes import guard_size_oblivious
+
+    if guard_size_oblivious(x.numel() == 0):
```
Comment (ColinPeppler) [inline torch/_decomp/decompositions.py:3931]: `GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(u0, 0) (unhinted: Eq(u0, 0)).  (Size-like symbols: u0)`

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -3428,7 +3428,8 @@ def inner_fn(idx):
             mask_val = ops.to_dtype(mask_loader(idx), torch.bool)
         else:
             mask_val = mask_loader(idx)
-        return ops.masked(mask_val, lambda: self_loader(_unsafe_index_fn(idx)), fill)
+        unsafe_idx = _unsafe_index_fn(idx)
+        return ops.masked(mask_val, lambda: self_loader(unsafe_idx), fill)
```
Comment (ColinPeppler) [inline torch/_inductor/lowering.py:3432]: Fixes this
```
# error - UnboundLocalError: local variable 'u0' referenced before assignment
bufX = ...
triton_(... bufX, u0 ...)
u0 = bufX.item()
```

Changing this evaluate `unsafe_index_fn` earlier gives `FreeUnbackedSymbolsOpsHandler` a chance to capture the `ops.indirect_indexing` called within `_unsafe_index_fn`. Before the handler doesn't capture it, so we lose `u0` as a dependency.

**Comment by lezcano** [comment]
Peter and I no longer work on `torch.compile`

**Comment by ColinPeppler** [comment]
eagerly evaluating unsafe_index_fn is creating issues.

File: torch/_inductor/dependencies.py [hunk 3]
```diff
@@ -734,6 +734,15 @@ def reduction(
...
 
+    def masked(self, mask, body, other) -> None:
+        if callable(body):
+            # The body is typically a lambda which can make
+            # additional calls like ops.indirect_indexing.
+            try:
+                body()
+            except TypeError:
+                pass
+
```
Comment (ColinPeppler) [inline torch/_inductor/dependencies.py:745]: Examples:

- _unsafe_masked_index
https://github.com/pytorch/pytorch/blob/76b5ee1119358ebba6cec9a97bb2574642954c4d/torch/_inductor/lowering.py#L3475

- pad_adaptive_loader
https://github.com/pytorch/pytorch/blob/76b5ee1119358ebba6cec9a97bb2574642954c4d/torch/_inductor/lowering.py#L4263-L4267

File: torch/_inductor/dependencies.py [hunk 4]
```diff
@@ -734,6 +734,15 @@ def reduction(
...
         return (None,) * num_values if num_values > 1 else None
 
+    def masked(self, mask, body, other) -> None:
+        if callable(body):
+            # The body is typically a lambda which can make
+            # additional calls like ops.indirect_indexing.
+            try:
+                body()
+            except TypeError:
+                pass
```
Comment (jansel) [inline torch/_inductor/dependencies.py:744]: When are we getting a type error here?  Should this be an error?

File: torch/_inductor/dependencies.py [hunk 5]
```diff
@@ -734,6 +734,15 @@ def reduction(
         num_values = reduction_num_outputs(reduction_type)
         return (None,) * num_values if num_values > 1 else None
 
+    def masked(self, mask, body, other) -> None:
+        if callable(body):
```
Comment (jansel) [inline torch/_inductor/dependencies.py:738]: Body should always be a callable, we should error if it isn't.

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by ezyang — COMMENTED** [review]
Endorsing the size oblivious change. I'll let jansel handle the rest

File: torch/_inductor/dependencies.py (hunk 4 above)
Comment (ColinPeppler) [reply torch/_inductor/dependencies.py]: Yup, this should be an error.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**